metagraph/src/graph/alignment/aligner_seeder_methods.cpp

```diff
@@ -36,269 +50,157 @@ void ExactMapSeeder<NodeType>::initialize(std::string_view query, bool orientati
-                callback(std::move(alt_alignment));
-            }
-        }
-
-        callback(std::move(alignment));
-    });
-
-    const auto &query_nodes = base_seeder_->get_query_nodes();
-    if (query_nodes.back())
-        return;
-
-    const auto &config = get_config();
-    const auto &graph = dynamic_cast<const DBGSuccinct&>(get_graph());
-    size_t k = graph.get_k();
+    size_t k = dbg_succ_.get_k();
     bool orientation = this->get_orientation();
 
-    for (size_t i = query.size() - k + 1; i + config.min_seed_length <= query.size(); ++i) {
-        graph.call_nodes_with_suffix_matching_longest_prefix(
-            std::string_view(query.data() + i, query.size() - i),
-            [&](NodeType alt_node, size_t seed_length) {
-                assert(alt_node != DeBruijnGraph::npos);
-                assert(seed_length < k);
-                assert(i + seed_length + 1 == query.size()
-                    || graph.traverse(alt_node, query[i + seed_length])
-                        == DeBruijnGraph::npos);
-
-                std::string_view seed_seq(query.data() + i, seed_length);
-                DBGAlignerConfig::score_t match_score = config.match_score(seed_seq);
-
-                if (match_score <= config.min_cell_score)
-                    return;
+    auto call_suffix_seed = [&](size_t i, node_index alt_node, size_t seed_length) {
+        std::string_view seed_seq(query.data() + i, seed_length);
+        DBGAlignerConfig::score_t match_score = config.match_score(seed_seq);
 
-                Seed seed(seed_seq, { alt_node }, match_score, i,
-                          orientation, k - seed_length);
-
-                assert(seed.is_valid(graph, &config));
-                callback(std::move(seed));
-            },
-            config.min_seed_length,
-            config.max_num_seeds_per_locus
-        );
-    }
-}
-
-template <typename NodeType>
-void MEMSeeder<NodeType>::initialize(std::string_view query, bool orientation) {
-    ExactMapSeeder<NodeType>::initialize(query, orientation);
-
-    const auto &query_nodes = this->get_query_nodes();
-
-    query_node_flags_.assign(query_nodes.size(), 0);
+        if (match_score > config.min_cell_score) {
+            Seed seed(seed_seq,
+                      { alt_node },
+                      match_score,
+                      i,
+                      orientation,
+                      k - seed_length);
 
-    for (size_t i = 0; i < query_nodes.size(); ++i) {
-        if (query_nodes[i] != DeBruijnGraph::npos) {
-            query_node_flags_[i] = 2 | (*is_mem_terminus_)[query_nodes[i]];
+            assert(seed.is_valid(dbg_succ_, &config));
+            callback(std::move(seed));
+        }
+    };
+
+    for (size_t i = 0; i + config.min_seed_length <= query.size(); ++i) {
+        if (i >= query_nodes.size() || !query_nodes[i]) {
+            size_t max_seed_length = std::min(config.max_seed_length,
+                std::min(k, query.size() - i)
+            );
```

Comment:
if you have many statements like this, it would make sense to write a new min in utils.

`utils::min(first, second, third, ...);`

Reply:
Turns out `std::min` works with initializer lists...